### ISSUE TITLE
P1-23: psn train を build_psn_training factory に分離する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#515]
---
title: "P1-23: psn train を build_psn_training factory に分離する"
labels: ["refactor", "P1", "codex"]
---

# P1-23: psn train を build_psn_training factory に分離する

## 依存 issue

21, 22

## 対象ファイル

- `packages/seisai-engine/src/seisai_engine/pipelines/psn/train.py`
- `packages/seisai-engine/src/seisai_engine/pipelines/psn/training.py（新規）`
- `cli/run_psn_train.py`
- `packages/seisai-engine/tests/test_psn*.py`
- `packages/seisai-engine/tests/test_pair_psn_loss_config.py`

## 背景

`psn/train.py` は noise mix、train-batch-only dataset、primary key validation、phase visualization などが main 近辺にあり、pair より複雑。pair factory 移行後のパターンを適用する。

## 目的

`build_psn_training(cfg, base_dir, ...) -> PipelineBuildResult` を作り、psn main を薄くする。

## 非対象

- noise mix の挙動変更はしない。
- phase pick dataset / loss の仕様変更はしない。
- CLI 引数は変更しない。

## 実装手順

1. `psn/training.py` に dataset build、noise mix、model/loss/optimizer build、infer_epoch_fn 構築を移す。
2. `_parse_noise_segy_files`, `_build_noise_detect_cfg`, `_build_noise_sampler_cfg`, `_maybe_wrap_train_dataset_with_noise_mix` などは `training.py` へ移すか、noise 専用 module に分ける。まずは `training.py` でよい。
3. `psn/train.py` は config 読込、listfile expansion、typed config load、factory call、spec 作成、runner call に限定する。
4. pair と同じ common training assembly helper を使う。
5. tests は psn 固有挙動を package factory の public API に寄せる。

## 受け入れ条件

- `psn/train.py` が orchestration 中心に縮小される。
- `build_psn_training` が単体 import 可能。
- noise mix と psn loss 関連 tests が通る。
- pair/psn で training assembly の利用パターンが揃っている。

## 推奨テスト

```bash
PYTHONPATH=packages/seisai-engine/src:packages/seisai-utils/src:packages/seisai-dataset/src:packages/seisai-transforms/src:packages/seisai-models/src:packages/seisai-pick/src \
python -m pytest \
  packages/seisai-engine/tests/test_psn_noise_mix_config_smoke.py \
  packages/seisai-engine/tests/test_psn_mix_with_noise_dataset.py \
  packages/seisai-engine/tests/test_psn_loss_label_valid_mask.py \
  packages/seisai-engine/tests/test_pair_psn_loss_config.py
```

```bash
PYTHONPATH=packages/seisai-engine/src:packages/seisai-utils/src:packages/seisai-dataset/src:packages/seisai-transforms/src:packages/seisai-models/src:packages/seisai-pick/src \
python -m pytest packages/seisai-engine/tests/test_import_smoke.py packages/seisai-engine/tests/test_lazy_import_boundaries.py
```
